src/network.rs

```diff
@@ -0,0 +1,124 @@
+use crate::{
+    member::{AlertMessage, UnitMessage},
+    nodes::NodeIndex,
+    Data, Hasher, Receiver, Sender,
+};
+use codec::{Decode, Encode};
+use futures::{channel::oneshot, FutureExt, StreamExt};
+use log::error;
+use std::fmt::Debug;
+
+#[async_trait::async_trait]
+pub trait Network: Send {
+    type Error: Debug;
+    fn send(&self, data: &[u8], node: NodeIndex) -> Result<(), Self::Error>;
+    fn broadcast(&self, data: &[u8]) -> Result<(), Self::Error>;
+    async fn next_event(&mut self) -> Option<Vec<u8>>;
+}
+
+#[derive(Encode, Decode)]
+enum NetworkData<H: Hasher, D: Data, S: Encode + Decode> {
+    Units(UnitMessage<H, D, S>),
+    Alert(AlertMessage<H, D, S>),
+}
+
+pub enum Recipient {
+    Everyone,
+    Node(NodeIndex),
+}
+
+pub(crate) struct NetworkHub<N: Network, H: Hasher, D: Data, S: Encode + Decode> {
+    network: N,
+    outgoing_units: Receiver<(UnitMessage<H, D, S>, Recipient)>,
```

Comment:
Is it just me, or is calling a receiver an output (as in `outgoing`) counterintuitive?

Reply:
Yup, very good point. These are outgoing units from the perspective of Rush as a whole, thus the name here, but I agree it's not good. How about `_to_send` and `_received` (with `units` and `alerts` as prefixes ofc)?

Reply:
Might be better now.

Reply:
Perfect!